pkg/module/download.go

```diff
@@ -0,0 +1,137 @@
+package module
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/gomods/athens/pkg/config"
+	"github.com/gomods/athens/pkg/storage"
+	multierror "github.com/hashicorp/go-multierror"
+)
+
+// Downloader downloads a module version from a URL exposing Download Protocol endpoints
+type Downloader func(ctx context.Context, timeout time.Duration, baseURL, module, version string) (*storage.Version, error)
+
+// Download downloads the module/version from url. Returns a storage.Version
+// representing the downloaded module/version or a non-nil error if something went wrong
+func Download(ctx context.Context, timeout time.Duration, baseURL, module, version string) (*storage.Version, error) {
+	tctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	var info []byte
+	var infoErr error
+
+	var mod []byte
+	var modErr error
+
+	var zip io.ReadCloser
+	var zipErr error
+
+	var wg sync.WaitGroup
+	wg.Add(3)
+
+	go func() {
+		defer wg.Done()
+		infoReq, err := getRequest(tctx, baseURL, module, version, ".info")
+		if err != nil {
+			info, infoErr = nil, err
+			return
+		}
+		infoStream, err := getResBody(infoReq, timeout)
+		if err != nil {
+			info, infoErr = nil, err
+			return
+		}
+		info, infoErr = getBytes(infoStream)
+	}()
+
+	go func() {
+		defer wg.Done()
+		modReq, err := getRequest(tctx, baseURL, module, version, ".mod")
+		if err != nil {
+			mod, modErr = nil, err
+			return
+		}
+		modStream, err := getResBody(modReq, timeout)
+		if err != nil {
+			mod, modErr = nil, err
+			return
+		}
+		mod, modErr = getBytes(modStream)
+	}()
+
+	go func() {
+		defer wg.Done()
+		zipReq, err := getRequest(tctx, baseURL, module, version, ".zip")
+		if err != nil {
+			zip, zipErr = nil, err
+			return
+		}
+		zip, zipErr = getResBody(zipReq, timeout)
+	}()
+	wg.Wait()
+
+	var errors error
+	if infoErr != nil {
```

Comment:
I think you can append the errors during the check inside the anonymous funcs for each file, then just check once here if `errors` is not nil and return it if needed. It would save the extra checks for each value

Reply:
@robjloranger is `multierror.Append` thread-safe?

Reply:
Good call, I don't think so.